File: gateway/gateway.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <pthread.h>

#define DEFAULT_GATEWAY_PORT 9010
#define MAX_VEHICLES 64
#define BUFFER_SIZE 1024
#define ID_MAX_LEN 64

typedef struct {
    int socket;
    char id[ID_MAX_LEN];
    struct sockaddr_in addr;
    int connected;
} VehicleEntry;

VehicleEntry vehicle_table[MAX_VEHICLES];
pthread_mutex_t table_lock = PTHREAD_MUTEX_INITIALIZER;

int get_gateway_port() {
    const char* env = getenv("GATEWAY_PORT");
    return env ? atoi(env) : DEFAULT_GATEWAY_PORT;
}

void print_vehicle_table() {
    printf("\n[INFO] Vehicle table:\n");
    pthread_mutex_lock(&table_lock);
    for (int i = 0; i < MAX_VEHICLES; i++) {
        if (vehicle_table[i].connected) {
            printf("  - %s : Connected (%s)\n",
                   vehicle_table[i].id,
                   inet_ntoa(vehicle_table[i].addr.sin_addr));
        }
    }
    pthread_mutex_unlock(&table_lock);
    printf("\n");
}

int register_vehicle(int client_fd, struct sockaddr_in client_addr, const char* id) {
    pthread_mutex_lock(&table_lock);
    for (int i = 0; i < MAX_VEHICLES; i++) {
        if (!vehicle_table[i].connected) {
            vehicle_table[i].socket = client_fd;
            strncpy(vehicle_table[i].id, id, ID_MAX_LEN - 1);
            vehicle_table[i].addr = client_addr;
            vehicle_table[i].connected = 1;
            pthread_mutex_unlock(&table_lock);
            return i;
        }
    }
    pthread_mutex_unlock(&table_lock);
    return -1;
}

void unregister_vehicle(int index) {
    pthread_mutex_lock(&table_lock);
    if (vehicle_table[index].connected) {
        close(vehicle_table[index].socket);
        vehicle_table[index].connected = 0;
        printf("[INFO] Disconnected: %s\n", vehicle_table[index].id);
    }
    pthread_mutex_unlock(&table_lock);
    print_vehicle_table();
}

void* handle_vehicle(void* arg) {
    int client_fd = *(int*)arg;
    free(arg);
    struct sockaddr_in client_addr;
    socklen_t addr_len = sizeof(client_addr);
    getpeername(client_fd, (struct sockaddr*)&client_addr, &addr_len);

    char buffer[BUFFER_SIZE];
    int index = -1;
    char vehicle_id[ID_MAX_LEN] = "UNKNOWN";

    int bytes = recv(client_fd, buffer, BUFFER_SIZE - 1, 0);
    if (bytes <= 0) {
        close(client_fd);
        return NULL;
    }
    buffer[bytes] = '\0';

    if (strncmp(buffer, "vehicle_id:", 11) == 0) {
        strncpy(vehicle_id, buffer + 12, ID_MAX_LEN - 1);
        vehicle_id[strcspn(vehicle_id, "\r\n")] = 0;  // strip newline
        index = register_vehicle(client_fd, client_addr, vehicle_id);
        printf("[INFO] Connected: %s (%s)\n", vehicle_id, inet_ntoa(client_addr.sin_addr));
        print_vehicle_table();
    } else {
        printf("[WARN] No vehicle_id received from %s\n", inet_ntoa(client_addr.sin_addr));
        close(client_fd);
        return NULL;
    }

    while (1) {
        bytes = recv(client_fd, buffer, BUFFER_SIZE - 1, 0);
        if (bytes <= 0) break;
        buffer[bytes] = '\0';
        printf("[%s] %s", vehicle_id, buffer);
        fflush(stdout);
    }

    if (index >= 0) {
        unregister_vehicle(index);
    } else {
        close(client_fd);
    }

    return NULL;
}

int main() {
    int server_fd, client_fd;
    struct sockaddr_in server_addr, client_addr;
    socklen_t addrlen = sizeof(client_addr);

    int port = get_gateway_port();

    server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0) { perror("socket failed"); exit(1); }

    int opt = 1;
    setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = INADDR_ANY;
    server_addr.sin_port = htons(port);

    if (bind(server_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("bind failed");
        exit(1);
    }

    listen(server_fd, 5);
    printf("[GATEWAY] Listening for vehicles on port %d...\n", port);

    while (1) {
        client_fd = accept(server_fd, (struct sockaddr*)&client_addr, &addrlen);
        if (client_fd < 0) {
            perror("accept failed");
            continue;
        }

        int* pclient = malloc(sizeof(int));
        *pclient = client_fd;
        pthread_t tid;
        pthread_create(&tid, NULL, handle_vehicle, pclient);
        pthread_detach(tid);
    }

    close(server_fd);
    return 0;
}

File: gateway/gateway.py
import socket
import threading
import os
import json
import time
from datetime import datetime
import paho.mqtt.client as mqtt

DEFAULT_TCP_PORT = 9010
DEFAULT_MQTT_HOST = "mqtt"
DEFAULT_MQTT_PORT = 1883
BUFFER_SIZE = 1024

vehicle_table = {}
table_lock = threading.Lock()


def log(msg):
    now = datetime.now().strftime("%H:%M:%S")
    print(f"[{now}] {msg}", flush=True)


def print_vehicle_table():
    log("Vehicle table:")
    with table_lock:
        for vehicle_id, info in vehicle_table.items():
            status = "Connected" if info["connected"] else "Disconnected"
            ip = info["ip"]
            log(f"  - {vehicle_id}: {status} ({ip})")


mqtt_host = os.environ.get("MQTT_BROKER_HOST", DEFAULT_MQTT_HOST)
mqtt_port = int(os.environ.get("MQTT_BROKER_PORT", DEFAULT_MQTT_PORT))

mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)


def handle_client(conn, addr):
    vehicle_id = None
    try:
        data = conn.recv(BUFFER_SIZE).decode().strip()
        if not data.startswith("vehicle_id:"):
            log(f"Rejected connection from {addr[0]}: no vehicle_id")
            conn.close()
            return

        vehicle_id = data.split("\n", 1)[0].strip()
        with table_lock:
            vehicle_table[vehicle_id] = {
                "ip": addr[0],
                "connected": True,
                "socket": conn,
            }

        log(f"Connected: {vehicle_id} ({addr[0]})")
        print_vehicle_table()

        while True:
            data = conn.recv(BUFFER_SIZE)
            if not data:
                break
            payload = data.decode().strip()
            log(f"[{vehicle_id}] {payload}")

            try:
                parsed = json.loads(payload)
                for key, value in parsed.items():
                    topic = f"vehicles/{vehicle_id}/sensor/{key}"
                    mqtt_client.publish(topic, str(value))
            except json.JSONDecodeError:
                log(f"[WARN] Invalid JSON from {vehicle_id}")

    except Exception as e:
        log(f"[ERROR] Client {addr[0]}: {e}")
    finally:
        if vehicle_id:
            with table_lock:
                if vehicle_id in vehicle_table:
                    vehicle_table[vehicle_id]["connected"] = False
            log(f"Disconnected: {vehicle_id}")
            print_vehicle_table()
        conn.close()


def start_server(port):
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind(("0.0.0.0", port))
    server_socket.listen()
    log(f"Gateway listening on TCP port {port}...")

    while True:
        conn, addr = server_socket.accept()
        threading.Thread(target=handle_client, args=(
            conn, addr), daemon=True).start()


if __name__ == "__main__":
    mqtt_client.connect(mqtt_host, mqtt_port)
    mqtt_client.loop_start()
    tcp_port = int(os.environ.get("GATEWAY_PORT", DEFAULT_TCP_PORT))
    start_server(tcp_port)
